src/Microsoft.AspNet.Mvc.Razor/Compilation/RoslynCompilationService.cs

```diff
@@ -131,6 +125,56 @@ public CompilationResult Compile([NotNull] RelativeFileInfo fileInfo, [NotNull]
             }
         }
 
+        // Internal for unit testing
+        internal CompilationResult GetCompilationFailedResult(
+            string relativePath,
+            string compilationContent,
+            string assemblyName,
+            IEnumerable<Diagnostic> diagnostics)
+        {
+            var diagnosticGroups = diagnostics
+                .Where(IsError)
+                .GroupBy(diagnostic => GetFilePath(relativePath, diagnostic), StringComparer.Ordinal);
+
+            var failures = new List<ICompilationFailure>();
+            foreach (var group in diagnosticGroups)
+            {
+                var sourceFilePath = group.Key;
+                string sourceFileContent;
+                if (string.Equals(assemblyName, sourceFilePath, StringComparison.Ordinal))
```

Comment:
Yes, this is for generated code that we produce but don't map back to the cshtml - this includes things like default inherited directives (`IHtmlHelper<TModel>`) and any code that relies on a property \ method being present on the base type `DefineSection`, `RenderAsync` etc). 

* I have a tracking bug for the first case - https://github.com/aspnet/Mvc/issues/2373.
* I wasn't able to find the bug the last time I looked - https://github.com/aspnet/Mvc/issues/1968. Mapping to the generated code should give you more more useful information for errors of this sort.